Adding PHP 8.2 support

gxgpet opened the pull request:
This adds support for PHP 8.2.

Basically, fixing dynamic props assigning.

gxgpet committed:
Adding PHP 8.2 support

gxgpet committed:
Merge branch 'bcit-ci:develop' into develop

Khuthaily commented:
Thank you, @gxgpet. In addition to the changes your are proposing, I had to add `#[AllowDynamicProperties]` in `system > core > Controller.php` and `system > core > Router.php` in one of my projects to provide support for PHP 8.2.

gxgpet committed:
Adds $uri prop in Router, and AllowDynamicProps attribute for CI_Controller

gxgpet commented:
Thanks for your suggestion, @Khuthaily! I added the attribute only for `CI_Controller`; as for the router, the `uri` property was missing only, and I added it. I guess you didn't find something different for the router, right?

sneakyimp commented:
Where do we stand on 8.2 compatibility? Issue #6192 seems to basically be a duplicate of this issue and I'm pretty eager to apply this fix to a production system I'm upgrading to PHP 8.2. Are we sure it covers all the dynamic class assignments? Are there any unit tests dedicated to checking this? Has anyone used a static analysis tool to check for undeclared properties? I've checked out PHPStan expressly to find issues like this and it apparently [doesn't detect undefined property assignments on objects instantiated via a factory method](https://github.com/phpstan/phpstan/issues/8845).

sneakyimp commented:
In the interest of furthering the effort for 8.2 compatibility, I have downloaded the latest CI 3.1.13 and run php stan on it:
```bash
vendor/bin/phpstan analyse -l 2 codeigniter > phpstan-output.txt
```
This yields about 2090 errors (see attached [phpstan-output.tgz](https://github.com/bcit-ci/CodeIgniter/files/10679709/phpstan-output.tgz)), including 688 'Access to an undefined property' errors. One example of which is:

>Access to an undefined property CI_Migration::$lang.

I grepped those lines and wrote a quick PHP script to list all the distinct classes before the '::' and this is the result:
```text
$this(CI_DB_cubrid_driver)
$this(CI_DB_mssql_driver)
$this(CI_DB_mysql_driver)
$this(CI_DB_mysqli_driver)
$this(CI_DB_oci8_driver)
$this(CI_DB_pdo_cubrid_driver)
$this(CI_DB_pdo_dblib_driver)
$this(CI_DB_pdo_ibm_driver)
$this(CI_DB_pdo_mysql_driver)
$this(CI_DB_pdo_odbc_driver)
$this(CI_DB_pdo_sqlsrv_driver)
$this(CI_DB_postgre_driver)
$this(CI_DB_sqlsrv_driver)
CI_Controller
CI_DB_cubrid_driver
CI_DB_driver
CI_DB_ibase_driver
CI_DB_ibase_forge
CI_DB_mssql_driver
CI_DB_mysql_driver
CI_DB_mysqli_driver
CI_DB_oci8_driver
CI_DB_pdo_4d_driver
CI_DB_pdo_cubrid_driver
CI_DB_pdo_dblib_driver
CI_DB_pdo_driver
CI_DB_pdo_firebird_driver
CI_DB_pdo_firebird_forge
CI_DB_pdo_ibm_driver
CI_DB_pdo_informix_driver
CI_DB_pdo_mysql_driver
CI_DB_pdo_oci_driver
CI_DB_pdo_odbc_driver
CI_DB_pdo_pgsql_driver
CI_DB_pdo_pgsql_forge
CI_DB_pdo_result
CI_DB_pdo_sqlite_driver
CI_DB_pdo_sqlsrv_driver
CI_DB_postgre_driver
CI_DB_postgre_forge
CI_DB_sqlite3_driver
CI_DB_sqlite_driver
CI_DB_sqlsrv_driver
CI_Image_lib
CI_Input
CI_Javascript
CI_Jquery
CI_Migration
CI_Router
CI_Table
CI_URI
object
```
 
I am not at all certain, but it looks like we might have a bit more work to do if we want to insure CI3 is really ready for PHP 8.2.

Anyone have thoughts on this?

EDIT: It looks like PHPStan doesn't see the CI_DB class definition because CI3 defines this class in a conditional block. I think we might be able to ignore all the CI_DB_* classes because they are either extending `CI_DB_query_builder` or `CI_DB_driver` -- surely we can put the attribute in those files abstract parent classes, right?

sneakyimp commented:
Is there any progress on supporting PHP 8.2? I've seen a lot of people posting issues about it but they are all being closed except for this one, which seems to have the most up-to-date information and pull request. I badly need to update a server running a CI3 website and I'd very much like to have the site work with PHP 8.2 so I can avoid yet another later migration.

I don't see any declaration of AllowDynamicProperties in this pull request before the critical CI_DB declaration [here](https://github.com/gxgpet/CodeIgniter/blob/develop/system/database/DB.php#L163). Nor do I see one applied to the CI_DB_query_builder abstract class [here](https://github.com/gxgpet/CodeIgniter/blob/develop/system/database/DB_query_builder.php#L53). I don't actually know if you can put AllowDynamicProperties in front of an abstract class, but it seems to me that we might fix most of those CI_DB_x_driver classes reported by phpstan if we were to address those two class declarations.

CI_Controller already has the dynamic properties attribute declared [here](https://github.com/gxgpet/CodeIgniter/blob/develop/system/core/Controller.php#L53) in this pull request.

That would leave the following classes reported by phpstan:
CI_DB_ibase_forge
CI_DB_pdo_firebird_forge
CI_DB_pdo_pgsql_forge
CI_DB_pdo_result
CI_DB_postgre_forge
CI_Image_lib
CI_Input
CI_Javascript
CI_Jquery
CI_Migration
CI_Router
CI_Table
CI_URI

Those forge and result classes could be handled, I think, by declaring AllowDynamicProperties for their parent classes.

Curiously, CI_Javascript and CI_Jquery classes do not exist in this pull request, so I'm not sure what to do about them.

The various other classes might also be solved with an AllowDynamicProperties declaration. Personally, I see no harm in simply adding these various declarations, and then it seems we are likely to have a viable PHP 8.2 fix, no?  

Do we have (or need) Unit tests to trigger the depreciated warning for these classes? What must we do to make progress on this issue?

sneakyimp commented:
It occurred to me to run phpstan on gxgpet's pull request instead of the CI3 codebase, and PHPstan is reporting a lot fewer errors. See attached [gxgpet-output.zip](https://github.com/bcit-ci/CodeIgniter/files/10788184/gxgpet-output.zip) which I generated like so:
```bash
vendor/bin/phpstan analyse -l 2 gxgpet > gxgpet-output.txt
```
This is still reporting about 1000 errors, but only 54 of those refer to undefined properties:
```text
$ grep 'undefined prop' gxgpet-output.txt
  410    Access to an undefined property object::$directory.
  412    Access to an undefined property object::$directory.
  281    Access to an undefined property CI_Input::$raw_input_stream.
  465    Access to an undefined property object::$dbdriver.
  465    Access to an undefined property object::$dbdriver.
  469    Access to an undefined property object::$dbdriver.
  182    Access to an undefined property object::$dbdriver.
  182    Access to an undefined property object::$dbdriver.
  187    Access to an undefined property object::$dbdriver.
  1480   Access to an undefined property CI_DB_driver::$qb_limit.
  474    Access to an undefined property object::$dbprefix.
  545    Access to an undefined property object::$dbprefix.
  548    Access to an undefined property object::$dbprefix.
  102    Access to an undefined property CI_DB_ibase_forge::$hostname.
  117    Access to an undefined property CI_DB_ibase_forge::$conn_id.
  123    Access to an undefined property object::$database.
  157    Access to an undefined property CI_DB_mysql_driver::$db.
  230    Access to an undefined property CI_DB_mysqli_driver::$db.
  146    Access to an undefined property CI_DB_pdo_result::$db.
  148    Access to an undefined property CI_DB_pdo_result::$db.
  88     Access to an undefined property CI_DB_pdo_firebird_forge::$hostname.
  103    Access to an undefined property CI_DB_pdo_firebird_forge::$conn_id.
  109    Access to an undefined property object::$database.
  102    Access to an undefined property CI_DB_pdo_pgsql_forge::$create_table_if.
  133    Access to an undefined property object::$database.
  162    Access to an undefined property CI_DB_postgre_driver::$db.
  90     Access to an undefined property CI_DB_postgre_forge::$create_table_if.
  119    Access to an undefined property object::$database.
  300    Access to an undefined property CI_DB_sqlsrv_driver::$_escape_like_chr.
  300    Access to an undefined property CI_DB_sqlsrv_driver::$_escape_like_str.
  558    Access to an undefined property CI_Image_lib::$dest_image.
  563    Access to an undefined property CI_Image_lib::$dest_image.
  571    Access to an undefined property CI_Image_lib::$dest_image.
  577    Access to an undefined property CI_Image_lib::$dest_image.
  145    Access to an undefined property CI_Migration::$lang.
  148    Access to an undefined property CI_Migration::$load.
  168    Access to an undefined property CI_Migration::$db.
  170    Access to an undefined property CI_Migration::$dbforge.
  174    Access to an undefined property CI_Migration::$dbforge.
  176    Access to an undefined property CI_Migration::$db.
  215    Access to an undefined property CI_Migration::$lang.
  276    Access to an undefined property CI_Migration::$lang.
  289    Access to an undefined property CI_Migration::$lang.
  294    Access to an undefined property CI_Migration::$lang.
  337    Access to an undefined property CI_Migration::$lang.
  396    Access to an undefined property CI_Migration::$lang.
  446    Access to an undefined property CI_Migration::$db.
  460    Access to an undefined property CI_Migration::$db.
  521    Access to an undefined property object::$lang.
  521    Access to an undefined property object::$lang.
  521    Access to an undefined property object::$lang.
  521    Access to an undefined property object::$lang.
  521    Access to an undefined property object::$lang.
  43     Access to an undefined property CI_TestCase::$ci_view_root.
```
I have started picking through some of these and believe that quiet a few might be resolved by declaring var types with php doc comments. For example, those first two complaints are in `system/core/CodeIgniter.php` because we don't declare a var type for the var `$RTR`:
```php
/*
 * ------------------------------------------------------
 *  Instantiate the routing class and set the routing
 * ------------------------------------------------------
 */
        $RTR =& load_class('Router', 'core', isset($routing) ? $routing : NULL);
```
Those first two PHPStan complaints could be resolved if we add an extra asterisk and `@var` declaration to the comment so it declares the var type:
```php
/**
 * ------------------------------------------------------
 *  Instantiate the routing class and set the routing
 * ------------------------------------------------------
 * @var CI_Router|NULL
 */
        $RTR =& load_class('Router', 'core', isset($routing) ? $routing : NULL);
```
We might similarly remedy the complaints about object::$dbdriver in system/core/Loader.php by declaring a type more specific than simply 'object.' 

Numerous complaints in Loader.php are due to the bad return type declaration in this function:
```php
        /**
         * CI Component getter
         *
         * Get a reference to a specific library or model.
         *
         * @param       string  $component      Component name
         * @return      bool
         */
        protected function &_ci_get_component($component)
        {
                $CI =& get_instance();
                return $CI->$component;
        }
```
This peculiar function looks suspiciously similar to a __get magic method but you'd have to expressly invoke it rather than relying on 'magic' invocation. A grep search of the code shows that this function is *only* invoked in `./system/core/Loader.php` to retrieve a config object. It is never used in any other file and never used to retrieve anything but a config object. We can rid ourselves of several PHPStan errors simply by changing the return declaration from `@return bool` to `@return object`.

I'd be happy to try and put in a little time to try and get rid of these 54 'undefined property' errors if folks think that would be a good idea. I would hate to go through the trouble only to have my efforts ignored.

Thoughts? Anyone?

steveadamsfixedit commented:
I'd very much like to see PHP 8.2 supported...

gxgpet committed:
Adding missing `dest_image` property in CI_Image_lib

gxgpet commented:
@sneakyimp 

> Are we sure it covers all the dynamic class assignments?

No, we are not, that's why I waited for more reports to come up... As part of CI 3's architecture, many of those can be found only if the code is actually executed.

> In the interest of furthering the effort for 8.2 compatibility, I have downloaded the latest CI 3.1.13 and run php stan on it:

This is indeed a very easy step to do, and I made it too with another tool; the problem is that the output of such a tool is containing a lot of false positives, and they must be sorted out... No such tool can understand properly how the code works, it's just static analysis.

But anyway, if anyone is concerned with the file you uploaded, here's the analysis I could made for it:

* We added `#[AllowDynamicProperties]` on `CI_DB_driver`, which is the root class for all implementations: `CI_DB_query_builder extends CI_DB_driver`, `CI_DB extends CI_DB_query_builder`. This can be easily found & proven if the code is opened with an IDE. So there's no issue on the DB layer and that only class which has the attribute is enough.
* We are not discussing `Access to an undefined property X`, `Method X() should return Y but return statement is missing`, `Constant X not found` and `Variable X might not be defined.`. These are obvious false positives, and they are not even related to the new PHP version.
* `PHPDoc tag X has invalid value` - nothing here
* `CI_Migration` has the `__get` magic method implemented
* `CI_Image_lib` - a commit was pushed with adding the missing property
* `CI_Input` has the `__get` magic method implemented.
* `CI_Javascript`, `CI_Jquery` - these were removed in 2016
* `CI_Router`, `CI_Table` and `CI_URI` are already changed here

Sorry, but any bug/issue we might have with PHP 8.2 must be reviewed and checked by a human eye to see if that's the case. We can't change the entire code just to avoid any false positives from a static analysis tool, we won't do it.

> I'd very much like to see PHP 8.2 supported...

Well, I guess we are at the end of it since nothing more popped up recently. I think we shall deploy a new version soon.

sneakyimp commented:
As I said last week:

>I'd be happy to try and put in a little time to try and get rid of these 54 'undefined property' errors if folks think that would be a good idea. I would hate to go through the trouble only to have my efforts ignored.

I went and examined *every* 'undefined property' complaint from PHPStan and fixed several bugs and hundreds of bad PHPDoc declarations in #6198. That PR contains every change you've made in this one, and numerous other improvements as well. I believe we should use that PR instead. Either that or merge its changes into this PR. If you don't like the change to the file system/database/DB.php, you can leave it out. Personally, I think it's weird and bad style to dynamically declare that trivial CI_DB class inside a conditional, which is inside a function. I think the reason for that kludge in the legacy CI3 code is an inept way of supporting unit testing, which is also poorly implemented in this project.

NielBuys commented:
Can somebody give me the "composer.json" changes/steps of how I can test this Pull request. I want to test and see if my app is compatible with this pull request.

NelsonXavier30 commented:
I couldn't figure out how to use my application without these errors

NelsonXavier30 commented:
someone to help me?

NielBuys commented:
I created a patch file from this pull request and merged the patch file into my fork of CodeIgniter to test my app.

I had 2 more errors I needed to fix in CodeIgniter before my app started working. I needed to add "AllowDynamicProperties" to Controller.php and Router.php. See attached patch file for my changes.

NOTE: I see some changes was not included in original patch file. The below changes is not needed to be added. I changed my branch to include all commits of this pull request.

[Add AllowDynamicProperties to Controller and Router.zip](https://github.com/bcit-ci/CodeIgniter/files/10980864/Add.AllowDynamicProperties.to.Controller.and.Router.zip)

I created a test environment for myself If somebody want to use it temporarily you are welcome. I can't guarantee the branch and I don't support it. Its only for myself.

What I needed to change in my app to use it.
1. Change "codeigniter/framework" to "nielbuys/framework"  in the "composer.json" file. 
2. Change in main "index.php" the following line from "$system_path = 'vendor/codeigniter/framework/system';" to "$system_path = 'vendor/nielbuys/framework/system';"
3. Then run composer update. Then composer will install the test framework v3.1.13.2

My app works in my test environment.

Here is the patch file I created from this pull request for record purposes.
[6173.zip](https://github.com/bcit-ci/CodeIgniter/files/10981083/6173.zip)


Thanks all

sneakyimp commented:
> I had 2 more errors I needed to fix in CodeIgniter before my app started working. I needed to add "AllowDynamicProperties" to Controller.php and Router.php. See attached patch file for my changes.

I think you may a little confused about which code you are working with. The pull request discussed in this issue is the one suggested by gxgpet, and [you can see](https://github.com/gxgpet/CodeIgniter/commit/e9ff2f2329244176c9275022c3cb5b3916123bcc#diff-0127b792d68f41aed61f1a8fb5456c7f59e149c0a558e268cc539aed878fa6fe) in that pull request that he has the AllowDynamicProperties declaration in [Controller](https://github.com/gxgpet/CodeIgniter/blob/develop/system/core/Controller.php#L53) and in [Router](https://github.com/gxgpet/CodeIgniter/blob/develop/system/core/Router.php) he added a property or two to hopefully resolve dynamic property complaints.

wagnerfnds commented:
Honestly it is sad to see codeigniter3 dying, just now that PHP is comin to be a very modern language. Time to migrate our projects i think.

RedDragonWebDesign commented:
What are next steps to get a code review and merge for this? Looks like it's been sitting for 4 months. Do we need to ping some additional collaborators? @gxgpet looks like you have merge access for this repo, any additional thoughts?

RedDragonWebDesign commented:
@narfbg

NielBuys commented:
> > I had 2 more errors I needed to fix in CodeIgniter before my app started working. I needed to add "AllowDynamicProperties" to Controller.php and Router.php. See attached patch file for my changes.
> 
> I think you may a little confused about which code you are working with. The pull request discussed in this issue is the one suggested by gxgpet, and [you can see](https://github.com/gxgpet/CodeIgniter/commit/e9ff2f2329244176c9275022c3cb5b3916123bcc#diff-0127b792d68f41aed61f1a8fb5456c7f59e149c0a558e268cc539aed878fa6fe) in that pull request that he has the AllowDynamicProperties declaration in [Controller](https://github.com/gxgpet/CodeIgniter/blob/develop/system/core/Controller.php#L53) and in [Router](https://github.com/gxgpet/CodeIgniter/blob/develop/system/core/Router.php) he added a property or two to hopefully resolve dynamic property complaints.

Thanks for letting me know. I have realized after I created the post, that the patch file from Github did not include those 2 changes. I added a note to my original post that I have changed my branch to exactly match what is in this pull request.

colinalbion commented:
We have substantial CI3 applications running nicely on php 7.4  that we don't want to rewrite for CI4 just to be PHP 8x and 9.x compatible.  We need a version of CI3 that is compatible with 7.4 and above (ideally php8.1 and eventually php9).  I'd happily support an initiative to do this.

To rewrite for CI4 would involve upheaval to 20k+ lines of code whereas making CI 3 compatible with PHP 8+ would perhaps be just a few hundred of lines of code of change.   

Anyone else in the same position?

Edit:  After a few hours....added upgraded to 3.1.13, then added #[AllowDynamicProperties] to a few system files (Controler, URI etc) and a few of our own libs, we now have our app working with PHP8.2 (very pleased).  I recognise that this is not the final solution, but it is reassuring that there is a path forward where we don't have to entirely rewrite tens of thousands of CI3 code into CI4 or Laraval.  Again, it would be great to see a CI3 version that is compatible with PHP 8+ (perhaps dropping php5 support).

codestellar commented:
Will this branch be merged? I am looking forward to see php 8.2 support for CI3.

mohammadhzp commented:
@narfbg Are you able to merge this pull request?

nillander commented:
Eu tentei atualizar o meu CI para trabalhar com o PHP8.2 mas não consegui, o melhor que consegui foi o PHP8.1

nillander commented:
> Temos aplicativos CI3 substanciais rodando bem no php 7.4 que não queremos reescrever para CI4 apenas para serem compatíveis com PHP 8x e 9.x. Precisamos de uma versão do CI3 que seja compatível com 7.4 e superior (idealmente php8.1 e eventualmente php9). Eu ficaria feliz em apoiar uma iniciativa para fazer isso.
> 
> Reescrever para o CI4 envolveria a alteração de mais de 20 mil linhas de código, enquanto tornar o CI 3 compatível com o PHP 8+ seria talvez apenas algumas centenas de linhas de código de alteração.
> 
> Mais alguém na mesma posição?
> 
> Edit: Depois de algumas horas .... adicionado atualizado para 3.1.13, então adicionado #[AllowDynamicProperties] a alguns arquivos do sistema (Controlador, URI etc) e algumas de nossas próprias bibliotecas, agora temos nosso aplicativo funcionando com PHP8 .2 (muito satisfeito). Reconheço que esta não é a solução final, mas é reconfortante que haja um caminho a seguir onde não precisamos reescrever totalmente dezenas de milhares de código CI3 em CI4 ou Laraval. Novamente, seria ótimo ver uma versão do CI3 compatível com o PHP 8+ (talvez eliminando o suporte ao php5).

Ammgo, você poderia me ajudar a atualizar o meu CI para o php 8.2?

nillander commented:
> We have substantial CI3 applications running nicely on php 7.4 that we don't want to rewrite for CI4 just to be PHP 8x and 9.x compatible. We need a version of CI3 that is compatible with 7.4 and above (ideally php8.1 and eventually php9). I'd happily support an initiative to do this.
> 
> To rewrite for CI4 would involve upheaval to 20k+ lines of code whereas making CI 3 compatible with PHP 8+ would perhaps be just a few hundred of lines of code of change.
> 
> Anyone else in the same position?
> 
> Edit: After a few hours....added upgraded to 3.1.13, then added #[AllowDynamicProperties] to a few system files (Controler, URI etc) and a few of our own libs, we now have our app working with PHP8.2 (very pleased). I recognise that this is not the final solution, but it is reassuring that there is a path forward where we don't have to entirely rewrite tens of thousands of CI3 code into CI4 or Laraval. Again, it would be great to see a CI3 version that is compatible with PHP 8+ (perhaps dropping php5 support).

Amigo, você poderia dizer como fez para compatibilizar o CI3 com o PHP8.2?

thaibmt commented:
> Eu tentei atualizar o meu CI para trabalhar com o PHP8.2 mas não consegui, o melhor que consegui foi o PHP8.1

I can support you make CI support php 8.2. please contact me

nillander commented:
> > Eu tentei atualizar o meu CI para trabalhar com o PHP8.2 mas não consegui, o melhor que consegui foi o PHP8.1
> 
> I can support you make CI support php 8.2. please contact me

Did I contact you by email

thaibmt commented:
Yes. You can contact me by my email

nillander commented:
`<removed>`

totoprayogo1916 commented:
> @NielBuys
> Can somebody give me the "composer.json" changes/steps of how I can test this Pull request. I want to test and see if my app is compatible with this pull request.

```json

    "require": {

        "codeigniter/framework": "dev-develop"

     },
    "repositories": [
        {
            "type": "vcs",
            "url": "https://github.com/gxgpet/CodeIgniter.git"
        }
    ]

```

madwings commented:
Is anyone from the team interested in merging this pull request?

tenzap commented:
Another item that was probably missed:
```
An uncaught Exception was encountered

Type: ErrorException

Message: Creation of dynamic property CI_DB_postgre_forge::$create_table_if is deprecated

Filename: system/database/drivers/postgre/postgre_forge.php

Line Number: 90
```

carinihdev commented:
Can this branch be merged?

frost-byte commented:
What's the status of this? Is it going to be merged?

fernandoribeiro7 commented:
What is the expected release date for the merge?

mmihalev commented:
Any idea when this is going to be merged? Having support for php 8.2 is a must!

madwings commented:
> Any idea when this is going to be merged? Having support for php 8.2 is a must!

Most likely never or not in the near future. This project is dead. BCIT and CodeIgniter Foundation have focused on CI4 and CI3 is left to dust. Sad but this is the reality, they are not even releasing the project from their hands so it is transferred to someone else and continues the support. If you want PHP8.2 support - fork it and do it by yourself. This is the way now.

mmihalev commented:
@narfbg , Нарфе, ела го merge-ни бре...

pocketarc commented:
Honestly, that's what I did - I forked it, and been using it ever since. I'll maintain it for 8.3 as well. I'm not interested in dropping compatibility with anything or making any big changes, it's literally just to keep old projects running on an up-to-date system.

Edit: Feel free to use the fork I maintain, it's easy to set it up in composer.json as a custom repository, so it replaces the original codeigniter package. If you didn't install CI via composer.json, just grab the updated files as you normally do.

mohammadhzp commented:
> > Any idea when this is going to be merged? Having support for php 8.2 is a must!
> 
> Most likely never or not in the near future. This project is dead. BCIT and CodeIgniter Foundation have focused on CI4 and CI3 is left to dust. Sad but this is the reality, they are not even releasing the project from their hands so it is transferred to someone else and continues the support. If you want PHP8.2 support - fork it and do it by yourself. This is the way now.

As far as I can tell, It doesn't look dead.
- PHP 8.2 was officially released on December 8, 2022, this PR opened a month before the final stable release.
- CI3 is continuing what CI2 started, which means many projects built on CI2 over the years easily upgraded to CI3, but that's not the case with CI4. It's new, and new projects are using it only. That's an important point.
- CI3 is super stable, it's fast. even its security module is ported(Anti-XSS) and is popular.

RedDragonWebDesign commented:
@pocketarc - can you publish a release (https://github.com/pocketarc/codeigniter/releases) so that https://github.com/pocketarc/codeigniter is easier for me to install in composer? maybe also turn on the issues tab so I can file a ticket for this kind of thing in the future? thanks

gxgpet committed:
Fixes the usage of `_create_table_if` in Postgres forge class

kenjis commented:
Forking a project is generally a bad option. Many forked projects have failed in the past.

If you are willing to maintain the project, it seems to me that you should first find a way to become a maintainer.
See
- https://forum.codeigniter.com/thread-76836.html
- https://github.com/bcit-ci/CodeIgniter/issues/6012

pocketarc commented:
@RedDragonWebDesign Package is up; pocketarc/codeigniter. Issues are up, too; happy to discuss anything there.

@kenjis It'd be great to become a maintainer, but I don't have any existing contributions to CI's repo, and so understandably they'd have no reason to trust me as a maintainer. It's also not something I particularly care about. All I want is to make sure some old CI projects I'm involved with can keep running on up-to-date servers without problems.

ewwink commented:
I'm using XAMPP (windows), is anyone experiencing high CPU usage using `#[AllowDynamicProperties]`?

jamieburchell commented:
> Well, I guess we are at the end of it since nothing more popped up recently. I think we shall deploy a new version soon.

@gxgpet Please can you merge this and release a new version?

gxgpet committed:
Merge branch 'bcit-ci:develop' into develop

trulyfurqan commented:
## PR Analysis

- **Main theme:** Adding PHP 8.2 support in a PHP-based project
- **PR summary:** The PR adds PHP 8.2 support to the project's continuous integration pipeline and updates the codebase to use PHP 8.2 features, specifically the AllowDynamicProperties attribute.
- **Type of PR:** Enhancement
- **[Score ](https://docs.bito.ai)(0-100, higher is better):** 80
- **Relevant tests added:** Yes
- **[Estimated effort to review ](https://docs.bito.ai)(1-5, lower is better):** 2
The PR is straightforward, involving the addition of PHP 8.2 to the test matrix and the application of the AllowDynamicProperties attribute to classes.
## PR Feedback

- **General suggestions:** The PR successfully introduces PHP 8.2 support, which is crucial for future-proofing the application. However, the use of AllowDynamicProperties should be carefully reviewed to ensure it doesn't inadvertently allow for unsafe property assignments which could lead to security issues. It's recommended to add property type declarations wherever possible to leverage PHP 8.2's type system for safer code. Additionally, the change to the private property name in the postgre_forge.php file could potentially break compatibility if external code relies on the previous name. This should be documented in the PR if it's an intentional breaking change.
## Code feedback

## file: .github/workflows/test-phpunit.yml

- **Suggestions:**



 1. Ensure that the addition of PHP 8.2 to the test matrix is reflected in the documentation and that all dependencies are compatible with PHP 8.2. [important] **Relevant line**:In .github/workflows/test-phpunit.yml at line 15
 ```
+ php: [ '8.2', '8.1', '8.0', '7.4', '7.3', '7.2', '7.1', '7.0', '5.6', '5.5', '5.4']
 ```
## file: system/database/drivers/postgre/postgre_forge.php

- **Suggestions:**



 1. Confirm that renaming the create_table_if property to _create_table_if doesn't affect any subclasses or method calls that might be using the old property name. If this is a necessary change, consider deprecating the old property rather than renaming it outright to maintain backward compatibility. [important] **Relevant line**:In system/database/drivers/postgre/postgre_forge.php at line 87
 ```
+ $this->_create_table_if = 'CREATE TABLE IF NOT EXISTS';
 ```
## file: system/libraries/Table.php

- **Suggestions:**



 1. Avoid using local variables that shadow class properties, as this can lead to confusion and maintenance issues. If the intention is to copy the default template, consider using a more descriptive variable name. [medium] **Relevant line**:In system/libraries/Table.php at line 490
 ```
+		$temp = $this->_default_template();
 ```
## file: tests/codeigniter/libraries/Upload_test.php

- **Suggestions:**



 1. Reorganize the test to set up the expected data array before setting properties on the upload object to improve readability and maintain the structure of the test. [medium] **Relevant line**:In tests/codeigniter/libraries/Upload_test.php at line 80
 ```
+		$data['file_path'] = '/tmp/';
 ```
## file: tests/mocks/ci_testcase.php

- **Suggestions:**



 1. Verify that the AllowDynamicProperties attribute is correctly applied to mock objects and that it does not interfere with the testing framework's ability to mock or override properties for testing purposes. [important] **Relevant line**:In tests/mocks/ci_testcase.php at line 1
 ```
+#[AllowDynamicProperties]
 ```
## file: system/core/Controller.php

- **Suggestions:**



 1. Consider adding actual property declarations in the classes where AllowDynamicProperties is used to benefit from PHP 8.2's enhanced type system and reduce the risk of unintended property assignments. [important] **Relevant line**:In system/core/Controller.php at line 50
 ```
+#[AllowDynamicProperties]
 ```

rickhewes commented:
Would be great to see this merged @gxgpet !

:+1:

tenzap commented:
[ubuntu-24.04-feature-freeze](https://launchpad.net/ubuntu/+milestone/ubuntu-24.04-feature-freeze) which is the next LTS Ubuntu will be on 2024-02-29.

The changes of this PR were added to 3.1.13 package in debian/ubuntu package so that it works, but if you could release a new version working up to latest php version soon that would be nice and the package could be updated accordingly. But 4 weeks is short.

jamieburchell commented:
At this point the only option is to make your own fork and apply your own patches.

1stwebdesigns commented:
Is anything happening with this?  Looks like it is about ready?

Nedbup commented:
Is there a possibility of merging this in? I would like to have support for php 8.2

Nedbup commented:
> Is anything happening with this? Looks like it is about ready?

any chance of having this merged in soon?

kenjis commented:
There is a fork project:
https://packagist.org/packages/pocketarc/codeigniter

(Added)
Another fork:
https://github.com/ib3ltd/CodeIgniter/tree/master

jamieburchell commented:
> There is a fork project: https://packagist.org/packages/pocketarc/codeigniter

The releases from this fork appear to be based on the `develop` branch and from scanning the commits, include functionality changes. The `develop` branch is [quite a bit different to the last released version](https://github.com/bcit-ci/CodeIgniter/compare/3.1.13...develop). 

IMO it would be better to cherry-pick  and apply specific fixes to the latest actual release/master if you want to keep as close as possible to the last official 3.1.13 release.

kenjis commented:
Indeed it is.
But since it is a fork, the person who forks it should do as s/he likes.

jamieburchell commented:
> Indeed it is. But since it is a fork, the person who forks it should do as s/he likes.

Of course. But people should be aware that it's not purely a bug fix fork of the latest release.

gxgpet commented:
All contributions & ideas to this PR are welcome, but can you please stop polluting it with everything non-PHP 8.2 support-related? Every time coming here to review the PR again takes a lot of time...

jamieburchell commented:
> can you please stop polluting it with everything non-PHP 8.2 support-related?

It's hardly surprising. This PR is from 2022. I think people are bored waiting/unable to wait/have given up waiting for it to be merged or a decision made or frustrated by the lack of answers to "can we get this merged".

RedDragonWebDesign commented:
@gxgpet Thanks for writing this patch. Is there someone you can call in a favor from in this organization to get your PR reviewed? @narfbg maybe? Or maybe you can self-merge?

pocketarc commented:
@jamieburchell Note that this PR is -also- based on the 'develop' branch, so even if this PR gets merged eventually, you'll still need to move on from the 3.1.13 release.

One thing that I've tried to do with the fork as well is revert breaking changes made in 'develop', precisely so that it can work as a drop-in upgrade for 3.1. I'd suggest giving it a try - it *should* work without any changes being necessary. And if anything pops up that does break something for you, I'll happily take care of it.

jamieburchell commented:
> @jamieburchell Note that this PR is -also- based on the 'develop' branch

@pocketarc Sure, but it can be merged in to stable master/3.1.13 to form a new 3.1.14 release without the need to use the entire `develop` branch as a 3.2.x release. I believe this is how the 3.1.x releases are made, and why you don't see `develop` commits from 2015 in 3.1.x.

Ultimately everyone is free to fork and use whatever they like so for the company I work at which uses 3.1.13 I decided to [fork purely for 3.1.x bug fixes and no BC](https://github.com/ib3ltd/CodeIgniter/compare/3.1.13...ib3ltd:CodeIgniter:master). Like you, I'm not interested in new features just PHP 8.x support and beyond. I also don't want to introduce all the changes from `develop`. The projects I have also use DataMapper ORM and the database changes made in `develop` are not compatible with that. I could fix it, but I don't _need_ to.

If a 3.1.14 is ever released (doubtful) hopefully I won't need to maintain a bug fix branch but my hand has been forced to do so because of PHP EOL.

1stwebdesigns commented:
@jamieburchell could you explain briefly how to use this patch without forking, or using a fork, of the develop branch?

jamieburchell commented:
> @jamieburchell could you explain briefly how to use this patch without forking, or using a fork, of the develop branch?

@1stwebdesigns I ended up creating a fork of this repo purely to service my own needs. I merged the changes I wanted from the `develop` branch in to my `master` (which is basically 3.1.13). Any changes I wanted that weren't yet merged in to `develop` I created my own PR for and merged those in to my `master` too. I vetted the changes to ensure no `develop` things crept in, like for example the cookie helper which has different defaults for parameters. Then I pointed my `composer.json` at my own fork and used `dev-master` as the version to use.

mackieee commented:
Does this PR generally need wider user testing before being put live? Is that's what holding this great work back?

kenjis commented:
If you expect this to be merged, why don't you all test it yourself, review it and approve it?

manusfreedom commented:
For your information, I have corrected the following code :
* in `system\database\DB_driver.php` instead of using `#[\AllowDynamicProperties]` : 
```
	/**
	 * List of dynamic properties
	 *
	 * @var	mixed[]
	 */
	protected $_internal_properties	= array();
	public function __set(string $name, mixed $value): void
	{
		if(isset($this->$name))
		{
			$this->$name = $value;
		}
		else
		{
			$this->_internal_properties[$name] = $value;
		}
	}
	public function &__get($name)
	{
		if(isset($this->$name))
		{
			return $this->$name;
		}
		else
		{
			return $this->_internal_properties[$name];
		}
	}
	public function __isset($name)
	{
		return isset($this->$name) || isset($this->_internal_properties[$name]);
	}
	public function __unset($name)
	{
		if(isset($this->$name))
		{
			unset($this->$name);
		}
		else
		{
			unset($this->_internal_properties[$name]);
		}
	}
```
* to `system\database\drivers\mysqli\mysqli_driver.php`  line 128: 
```
		// Do we have a socket path?
		if ($this->hostname && $this->hostname[0] === '/')
		{
```
I don't know why, but in my projects, `#[\AllowDynamicProperties]` was not needed in `Loader.php`.
I tried to avoid `#[\AllowDynamicProperties]` which will be deprecated in PHP 9, as I understand it, but I can't avoid it in `Controller.php` because of the use of set by reference at the same time as the __set&__get magic.

pocketarc commented:
@manusfreedom Note: Dynamic properties are what's deprecated; the attribute is not deprecated, nor is there a plan to deprecate it at the moment. The attribute exists to tell PHP "yes, we're using dynamic properties on purpose", which makes perfect sense in the case of a legacy framework built around dynamic properties.

From the RFC (emphasis mine):

> The creation of dynamic properties on classes **that aren't marked with the #[AllowDynamicProperties] attribute** is deprecated in PHP 8.2 and becomes an Error exception in PHP 9.0.

Source: https://wiki.php.net/rfc/deprecate_dynamic_properties

mackieee commented:
Currently in the process of updating our site to PHP 8.2 - Certainly adding `#[\AllowDynamicProperties]` to almost all system Controllers was required (E.g. `CI_Benchmark`), more so than what the pull request suggested seemingly. Once I had done this, looks to be in good shape.

jamieburchell commented:
> Currently in the process of updating our site to PHP 8.2 - Certainly adding `#[\AllowDynamicProperties]` to almost all system Controllers was required (E.g. `CI_Benchmark`), more so than what the pull request suggested seemingly. Once I had done this, looks to be in good shape.

The `CI_Benchmark` controller on its own doesn't appear to use dynamic properties. Perhaps you have overridden it?

mackieee commented:
> > Currently in the process of updating our site to PHP 8.2 - Certainly adding `#[\AllowDynamicProperties]` to almost all system Controllers was required (E.g. `CI_Benchmark`), more so than what the pull request suggested seemingly. Once I had done this, looks to be in good shape.
> 
> The `CI_Benchmark` controller on its own doesn't appear to use dynamic properties. Perhaps you have overridden it? 

Hi Jamie, appears that may be the case (: 
I admit I've gone a little overzealous with adding the attribute where it may not be needed :)

Did you come across much else, in the way of issues in this pull request you forked from?

jamieburchell commented:
> Did you come across much else, in the way of issues in this pull request you forked from?

No issues with this PR. I have it merged in to a fork we are using in production along with some other fixes. You can see the difference between the fork and 3.1.13 [here](https://github.com/bcit-ci/CodeIgniter/compare/master...ib3ltd:CodeIgniter:master).

mackieee commented:
I have come across an issue in my tests - although I respect this is a little difficult to replicate due to the hardware stack, thus a bit of an edge case. It's one of that which `PHP >=8.1` has introduced.  

I've identified the issue but would appericate input if the fix could break other aspects of the query builder.

Our site uses Oracle OCI8 & MySQL DB instances. See below for a small example: 

``` php
$odbc  = $this->load->database( "oracle", true );
$mysql = $this->load->database( "mysql", true );

$products = $odbc->select([ "sku", "stock" ])->from( "products" )->get_compiled_select();
$users = $mysql->select([ "username", "dt_last_login" ])->from( "users" )->get_compiled_select();

echo PHP_VERSION . "<br />";
echo $products  . "<br />";
echo $users;
```

Running the above in `PHP <= 8.0` would yield: 

``` php
8.0.24
SELECT "sku", "stock" FROM "products"
SELECT `username`, `dt_last_login` FROM `users`
```

However in `PHP >= 8.1`: 

``` php 
8.1.28
SELECT "sku", "stock" FROM "products"
SELECT "username", "dt_last_login" FROM "users"
```
This produces a invalid MySQL query and thus a hard MySQL Error.  

Note the `escape_char` has been modified in the second query - note this only happens if a MySQL query attempts to be built *after* a connection that has a different modifier. 

The offending line is in `DB_Driver.php` - All PHP versions prior to 8.0 act in the same manner, so I'm assuming `static` has had some behaviour change in `PHP 8.1`.

https://github.com/bcit-ci/CodeIgniter/blob/6074d34d41046fe55754a231532042af49e1037e/system/database/DB_driver.php#L1417

Simply removing `static` does fix the issue, I'm assuming this was there originally as an optimisation. More than a functional use.

jamieburchell commented:
> I'm assuming `static` has had some behaviour change in `PHP 8.1`.

I think it's [this](https://wiki.php.net/rfc/static_variable_inheritance)

Not sure why that `preg_gc` couldn't be initialised when the class is instantiated and stored as its own class variable?

kenjis commented:
This might be useful for you. https://github.com/codeigniter4/CodeIgniter4/pull/5262

jamieburchell commented:
> This might be useful for you. [codeigniter4/CodeIgniter4#5262](https://github.com/codeigniter4/CodeIgniter4/pull/5262)

I'll open a PR for this later unless @gxgpet is planning on doing the same.

jamieburchell commented:
@mackieee Would you mind testing [this PR](https://github.com/bcit-ci/CodeIgniter/pull/6281) to see if this resolves your issue?

Please note that the PR is based from the develop branch and so merging only these changes in to the master/latest release needs to be done carefully (there will be some conflicts) so as not to copy over anything else that's changed. I'll merge this in to my fork if there are no issues found.

mackieee commented:
Wow, thanks for your time and effort here @jamieburchell 🥇 Highly appericated. I'll give this a test now :) 

Update: That worked a treat 👍 

As per your note on the usage of the static variables in the final classes, doesn't appear to have been removed in @kenjis's review on this in CI4 and is still current [repo](https://github.com/codeigniter4/CodeIgniter4/blob/f927928aaf42a0646301c629887d86ee913f84f2/system/Database/MySQLi/Result.php#L57). So I'd say we're safe there.

gordonb3 commented:
Came across this issue a while back but didn't have time to address it. The app I inherited was actually still running with CI2 and I didn't really see how to upgrade to CI4 which supposedly fixed the issue. Of course that turns out to be a no go completely but I now succeeded in getting it to the latest CI3 release. Definitely an improvement with respect to how the warnings are displayed in debug mode but the documentation on PHP appears to not be created for human reading and it took me at least an hour to conclude that you are supposed to declare a variable before using it.

```
Christ!
Why can't they just say that, instead of writing a ten page
essay showing examples of wrong code snippets?
```

My other understanding was that I might circumvent the issue by declaring the classes with `#[AllowDynamicProperties]` but I didn't really want to do that and eventually came to a similar result as you. I like how you formatted the changes to match the original code better, but there is one thing that I believe should not be in there:

You have added `#[AllowDynamicProperties]` to the `CI_Controller` class but there is really nothing in that class that would require this override. What I found is that whenever this class pops up a warning this is in fact caused by the extension class (i.e. your code) not having declared its additional variables. While this does allow old code to run unaltered with CI3 I do think this is not the right objective for the main code. The same may apply to other instances of this attribute which I can't verify because I don't use all of them. The only class I found where declaring `#[AllowDynamicProperties]` could not be avoided by making my own code compliant with PHP 8.2+ is `CI_Loader`.

jamieburchell commented:
> there is really nothing in that class that would require this override. 

Don't think that's correct. Example being when you load a library and reference it as `$this->foo` in the controller. I certainly ran in to warnings with the controller class and I always declare my own class properties.

gordonb3 commented:
> > there is really nothing in that class that would require this override.
> 
> Don't think that's correct. Example being when you load a library and reference it as `$this->foo` in the controller. I certainly ran in to warnings with the controller class and I always declare my own class properties.

That is what I initially thought as well, but on second view I noticed that the warnings still produced stated an origin address in `CI_Loader`.

jamieburchell commented:
@gordonb3 You can see why `#[AllowDynamicProperties]` is required for the controller [here](https://github.com/bcit-ci/CodeIgniter/blob/3658d731eaabe6117298a105ffb5b9dd59e190ce/system/core/Controller.php#L83)

gordonb3 commented:
> @gordonb3 You can see why `#[AllowDynamicProperties]` is required for the controller [here](https://github.com/bcit-ci/CodeIgniter/blob/3658d731eaabe6117298a105ffb5b9dd59e190ce/system/core/Controller.php#L83)

I don't have it and I'm not seeing any warnings.

What you need to do here is that if you want to use class Foo as `$this->foo` in your controller extension then you must declare $foo in the main body of your class. The line you referenced will then still generate a warning, however this warning does not come from controller but from CI_Loader (which appears to create a secondary reference as `$this->load->foo`). Try it, you'll see it works without making changes to CI_Controller itself.

jamieburchell commented:
> What you need to do here is that if you want to use class Foo as `$this->foo` in your controller extension then you must declare $foo in the main body of your class. 
 
You're free to do that, but I haven't encountered any CI projects where the libraries loaded are defined as controller class variables in that way. I'd say at the very least it wasn't typical for CI. By allowing dynamic variables, it provides PHP 8.2 compatibility for existing codebases that use CI libraries in the documented and typical way, without any effort.

gordonb3 commented:
> > What you need to do here is that if you want to use class Foo as `$this->foo` in your controller extension then you must declare $foo in the main body of your class.
> 
> You're free to do that, but I haven't encountered any CI projects where the libraries loaded are defined as controller class variables in that way. I'd say at the very least it wasn't typical for CI. By allowing dynamic variables, it provides PHP 8.2 compatibility for existing codebases that use CI libraries in the documented and typical way, without any effort.

That is what I said. Adding this attribute allows running old code that as of PHP 8.2+ is essentially bad. I understand how that is appealing but I don't think it is the responsibility of the core library to mask such bad code, nor to encourage people to create new bad code based on this functionality.

jamieburchell commented:
> That is what I said. Adding this attribute allows running old code that as of PHP 8.2+ is essentially bad. 

But this *is* how to do things in CI3. It's an ancient framework with its roots in ancient versions of PHP with compatibility maintained for those versions. There is nothing modern about CI3 in terms of PHP. That's what CI4 is for. These fixes are for desperate life support only, not for rethinking the framework.

It's also not really practical to define the libraries that might be loaded at any given time from any place in the application. Consider this code in some random library or helper somewhere:

```
$ci = & get_instance();
$ci->load->library('foo');
$ci->foo->bar(....);
```

gordonb3 commented:
> > That is what I said. Adding this attribute allows running old code that as of PHP 8.2+ is essentially bad.
> 
> But this _is_ how to do things in CI3. It's an ancient framework with its roots in ancient versions of PHP with compatibility maintained for those versions. There is nothing modern about CI3 in terms of PHP. That's what CI4 is for. These fixes are for desperate life support only, not for rethinking the framework.

If that is the objective you shouldn't upgrade PHP but run it in a hypervisor if hardware becomes an issue for the corresponding old OS

jamieburchell commented:
> If that is the objective you shouldn't upgrade PHP but run it in a hypervisor if hardware becomes an issue for the corresponding old OS

Or just add a single comment to the CI_Controller class to maintain compatibility with how the framework is intended to be used for PHP 8.2+ 🤷🏻‍♂️

gordonb3 commented:
> Or just add a single comment to the CI_Controller class to maintain compatibility with how the framework is intended to be used for PHP 8.2+ 🤷🏻‍♂️

Sure. You can do that. But on an individual base.

jamieburchell commented:
> Sure. You can do that. But on an individual base.

I disagree because this isn't a problem introduced by users writing sloppy code, rather a result of how the framework is structured to allow dynamic variables to be attached to the CI instance. And because the CI instance can be grabbed from any library or helper (or anywhere really) it's not feasible to pre-define every library that might be injected in to the CI instance. Much the same reason why this fix is required for the CI Loader. Of course, if the framework had done things differently in the first place perhaps by making use of a class array to hold these things, with magic getters and setters, dynamic properties wouldn't be needed at all. But like I say, this is to appease how the framework is designed without reengineering it.

gordonb3 commented:
It's not the CI Instance that made you create a class like

```
class Bar extends CI_Controller {

function Bar() {
      $this->foo = func(x);
}

}
```

In fact I'm sure that you already had to change `function Bar` to `function __construct` when moving to PHP 7 something.

jamieburchell commented:
I don't see what class constructors have to do with this issue. Simply put, the CI framework makes use of dynamic properties in CI_Controller (which is the CI super instance). You can see more examples [in the framework code](https://github.com/bcit-ci/CodeIgniter/blob/3658d731eaabe6117298a105ffb5b9dd59e190ce/system/core/Loader.php#L1302). Same as for CI_Loader

gordonb3 commented:
It's not about the class constructor. It's about referencing `$this-foo` without first declaring it, and thereby dynamically creating it, which is deprecated as of PHP 8.2.

jamieburchell commented:
[This has already been discussed](https://github.com/bcit-ci/CodeIgniter/issues/6192#issuecomment-1397748914). The fix decided upon is the use of the comment attribute in this PR.

It's almost certainly a moot point anyway; I don't believe there will ever be a merge of this to create a new release of CI3. Fork the framework and modify as you see fit, or use a fork that aligns better with your expectations. Better, if you care about best practice and modern PHP, use a modern framework. There are many, many aspects of CI3 that are no longer best practice in modern PHP - dynamically created class variables are only one example.

gordonb3 commented:
@jamieburchell  Might want to have a look how this PR addresses named issue with the very file mentioned in that link you provided.

umefarooq commented:
When this will merge in official base branch?

JaneX8 commented:
This should be merged.

Mhmad-Alnajjar commented:
Is there any estimated time for merging this pull request? i guess based on the above comments its crucial to have something official to support PHP 8.4

JaneX8 commented:
For future reference:
- https://github.com/pocketarc/codeigniter
- https://packagist.org/packages/pocketarc/codeigniter
